icu4j/main/core/src/main/java/com/ibm/icu/segmenter/BoundaryIteratorOfInts.java

```diff
@@ -0,0 +1,65 @@
+// © 2025 and later: Unicode, Inc. and others.
+// License & terms of use: https://www.unicode.org/copyright.html
+
+package com.ibm.icu.segmenter;
+
+import com.ibm.icu.text.BreakIterator;
+import com.ibm.icu.segmenter.Segments.IterationDirection;
+
+/**
+ * An iterator of segmentation boundaries that can operate in either the forwards or reverse
+ * direction.
+ *
+ * <p>When constructed to operate in the forwards direction, the iterator will return all boundaries
+ * that are strictly after the input index value provided to the constructor. However, when
+ * constructed to operate in the backwards direction, if the input index is already a segmentation
+ * boundary, then it will be included as the first value that the iterator returns as it iterates
+ * backwards.
+ */
+class BoundaryIteratorOfInts {
+  private BreakIterator breakIter;
+  private IterationDirection direction;
+  private int currIdx;
+
+  BoundaryIteratorOfInts(BreakIterator breakIter, CharSequence sourceSequence, IterationDirection direction, int startIdx) {
+    this.breakIter = breakIter;
+    this.direction = direction;
+
+    if (direction == IterationDirection.FORWARDS) {
+      currIdx = breakIter.following(startIdx);
+    } else {
+      assert direction == IterationDirection.BACKWARDS;
+
+      // When iterating backwards over boundaries, adjust the start index forwards by 1 to
+      // counteract the behavior from BreakIterator.preceding(), which we use to initialize the
+      // BreakIterator state, that always moves backwards by at least 1. We want to support an
+      // API that includes the input index when it is itself a boundary, unlike the behavior of
+      // BreakIterator.preceding().
+      int sourceLength = sourceSequence.length();
+      boolean isOnBoundary =
+          0 <= startIdx
+          && startIdx <= sourceLength
+          && breakIter.isBoundary(startIdx);
+      int backFromIdx = isOnBoundary ? startIdx + 1 : startIdx;
```

Comment:
if startIdx == 0 then call breakIter.first(); if startIdx == sourceLength then call breakIter.last()

we only need this dance when we are in the middle